### DISCUSSION TITLE
Wireable incorrectly hydrates array of objects into array of arrays when used with #[Modelable]

### DISCUSSION THREAD

**Discussion by AlexanderSuvorovN** [livewire/livewire#8817] [7. Bugs (archived)]
### Livewire version

v3.5.6

### Laravel version

11.21.0

### Which PHP version are you using?

PHP 8.3

### 1. Define a parent livewire component:

```
<?php

use Livewire\Volt\Component;
use ExPo\Bug\WireCompany;

new class extends Component 
{
    public $counter;
    public WireCompany $company;

    public function mount ()
    {
        $this->counter = 0;
        $this->company = new WireCompany([ 'name' => "Company_".rand(1, 1000) ]);
    }

    public function increment()
    {
        $this->counter++;
    }
}; ?>

<div>

    <!-- this counter is simply to cause "de/hydration" of bug.companyForm model -->
    <h2>Counter</h2>
    <div>{{ $counter }}</div>
    <x-secondary-button wire:click="increment">Increment</x-secondary-button>

    <!-- incorrect de/hydration of the model: instead of WireCompany instance it is hydrated as array -->
    <livewire:bug.companyForm wire:model="company" />

</div>
```

### 2. Define child component

```
<?php

use Livewire\Volt\Component;
use Livewire\Attributes\Modelable;

use ExPo\Bug\WireCompany;
use ExPo\Bug\WireEmployee;

new class extends Component 
{
    #[Modelable]
    public $company;

    public function mount()
    {
        // 
        // if $this->company is used only within the component then employees 
        // get correctly hydrated into WireEmployee instance
        // 
        // $this->company = $this->company ?? new Company([ 'name' => "Child_Component_Company_".rand(1000, 2000) ]);        
    }

    public function addEmployee()
    {
        $this->company->employees[] = new WireEmployee();
    }

    public function deleteEmployee($index)
    {
        unset($this->company->employees[$index]);
        $this->company->employees = array_values($this->company->employees);
    }

}; ?>

<div>
    <h2 class="mt-6">{{ $company->name }}</h2>
    <x-secondary-button wire:click="addEmployee" class="my-4">Add Employee</x-secondary-button>
    <div class="flex flex-col gap-2">
        @foreach ($company->employees as $key => $employee)
            <div wire:key={{ $key }} class="grid grid-cols-3 bg-white p-4 rounded shadow">                
                <div>employee</div>
                <div>{{ $employee->number }}</div>
                <div>
                    <x-secondary-button wire:click="deleteEmployee({{ $key }})">Delete</x-secondary-button>
                </div>
            </div>
        @endforeach
    </div>
</div>
```

### 3. Click on "Add Employee" button

This will put an instance of WireEmployee into the employees array.

### 4. Click on the "Increment" button

This will cause de/hydration of the $company object and array of employees will be incorrectly dehydrated into array of arrays and the companyForm component breaks.

### Repository or Wirebox URL

https://wirebox.app/b/xvo8q

### WireCompany class

```
<?php

namespace ExPo\Bug;

use Livewire\Wireable;

class WireCompany implements Wireable
{
    public $name;
    public $empoloyees;

    public function __construct($data)
    {
        $this->name = $data['name'];
        $this->employees = $data['employees'] ?? [];
    }

    public function toLivewire()
    {
        $data = [];
        $data['name'] = $this->name;
        $data['employees'] = $this->employees;
        return $data;
    }

    public static function fromLivewire($value)
    {
        $data = [];
        $data['name'] = $value['name'];
        $data['employees'] = $value['employees'];
        return new static($data);
    }
}
```

### WireEmployee class

```
<?php

namespace ExPo\Bug;

use Livewire\Wireable;

class WireEmployee implements Wireable
{
    public $number;

    public function __construct($data = [])
    {
        $this->number = $data['number'] ?? rand(2000, 3000);
    }
 
    public function toLivewire()
    {
        $data = [];
        $data['number'] = $this->number;
        return $data;
    }

    public static function fromLivewire($value)
    {
        $data = [];
        $data['number'] = $value['number'];
        return new static($data);
    }
}

```

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by osama300** [comment]
did you slove problem?

**Reply by AlexanderSuvorovN** [reply]
Hi, my solution (better to say workaround) was to abandon use of Wireable interface and use PHP arrays instead. But later on I migrated by app from Livewire to separate React frontend so can't comment now if the problem was resolved with new Laravel 12.

**Reply by osama300** [reply]
I found a solution with:

> public function toLivewire()
> {
>     return get_object_vars($this);
> }

Why did you join the enemy team with React? 😆

**Reply by AlexanderSuvorovN** [reply]
Livewire is great for relatively simple UI and I think it's capable to handle most of UI interactions that React offers. I just feel that with React it's more natural to have UI processing on frontend side instead of roundtrip to the server which eventually leads to hydration and dehydration problems for complex UI workflows (e.g. multi-step form) and sophisticated data models. Modern apps have very complex multi-component UIs these days and it requires a lot of gluing on the frontend side for things working together smoothly. Also most of the apps are multiplatform these days so I will need to use APIs for mobile app anyway. For this reason I'm not using .NET Blazer either. In a long run it will also be easier to update front-end without messing around with backend code. I'm still using Laravel as my backend engine, though :-)

**Reply by osama300** [reply]
I agree with you in the end—every tool works in its own way, and that's something we have to accept.
Now, I have built a **SaaS eCommerce platform** myself using **Laravel** and its amazing tools: **FilamentPHP** and **Livewire**.
It was a dream for me, and the result it's very good.
I structured the logic into separate files: **Repo, Services, and DTO**, meaning it can be reused in both the **API** and **UI** alike.
On the other hand, if I wanted to build it differently, it would require a team and a lot of time—**API, Frontend**, and so on…
What keeps **Laravel** alive is that it makes some dreams come true :)

**Reply by AlexanderSuvorovN** [reply]
Awesome! I'm glad to hear about your success! Yes, I'm thinking to reserve Livewire for the admin workflow: I want to remain up to date about this concept and where it is going. Admin is more for backoffice and doesn't require flawless & catchy UI so Livewire seems to be a perfect fit here. And it looks like FilamentPHP is what I'm looking for, thank you for sharing!

**Reply by osama300** [reply]
Yes, these are good tools that can be used to increase productivity.
I wish you all the best.
